### ISSUE TITLE
Feature request: use regex or some kind of pattern matching to recognize symbols

### ISSUE THREAD

**Issue by Delta-dev-99** [Durzn/BracketHighlighter#14]
I use this extension to highlight c++ code. The specific use case I'm interested in are the following.
When recognizing `#if` `#endif` pairs the '#' character may be separated by white-space and the symbols are not recognized.
Furthermore, there are multiple forms (`#if`,`#ifdef`,`#ifndef`) that can contain one another (as `#ifdef` contains `#if`). Recognizing both symbols seems to break highlighting from content inside the larger symbol. Recognizing only the small one results in partial highlighting.

**Comment by Durzn** [comment]
I'll look at it when I find some time in the future, since this is something that I use as well. 
For me, highlighting the "#if" was enough for now, but I can see how it might be annoying. 
For a current workaround (how I'm using it): Make a new pair of symbols for each indentation (e.g. `# if` and `# endif` or `....#if` and `....#endif` depending on which side you indent, where "." represents a whitespace, which github sadly does not show).
This will not make variants like #ifdef any better for now though.

**Comment by ktekinay** [comment]
I have a related issue in Elixir. The pairs are "do/end", but there is also a "do:" that doesn't have an "end". The plugin will still consider "do:" and find the next, unrelated "end".

Also, this code makes the plugin highlight the wrong block:

```
def something do
  pend = "" # it stops here
end
```

Ideally I'd have the option to specify the symbols as a regular expression, like:

```
{"open": "\\bdo\\b(?!:)", "close": "\\bend\\b", "regex": true}
```

The regex pattern would be case-sensitive by default, and quoted text would always be separately highlighted.

**Comment by Durzn** [comment]
Using regex for the search would be nice, but it would also require a complete re-write of the extension unfortunately.
It has grown considerably from what it was initially supposed to do (basically just match () and {} if you click on them).
Adding new features and not breaking other old features is a real pain since there are no tests for regression and the code base is honestly extremely hard to maintain (thanks past me).

The kind words of the community have been a good motivator, if this feature is requested by many I might consider doing it.
Otherwise the extension is not a priority of mine anymore, so I would only do this if there is enough motivation to do so, which could or couldn't happen.
Sorry that I cannot give you more hopeful news other than: maybe in the future.

**Comment by ktekinay** [comment]
Just take 5 or 10 minutes to do a complete top-down rewrite with tests.

Really appreciate this extension even if this issue is never addressed, thanks.

**Comment by Durzn** [comment]
@ktekinay You motivated me to tackle this problem and I'm currently working on it in this pull request: https://github.com/Durzn/BracketHighlighter/pull/31
If you want to have a preview, you can try installing the attached *.vsix file there, I also included an example how the new customSymbols variable looks like.
It still needs some finetuning which I will do in the coming days, however.

**Comment by ktekinay** [comment]
That's awesome, thanks!

I'm in the middle of heavy-workload deadline right now, but I'm going to keep on eye on your progress and will try it at some point.

**Comment by Durzn** [comment]
I released the extension with version 3.0.0 as a pre-release build, so feel free to simply download it from the official marketplace when you find the time to try it out.
I'll close it for now, but if you find any bugs or have additional requests, feel free to open another issue.
